### PR TITLE
[RISC-V] Base scheduling model for tt-ascalon-d8

### PR DESCRIPTION
**Pull request by ppenzin**
First part of tt-ascalon-d8 scheduling model, only containing scalar ops. Scheduling for vector instructions will be added in a follow-up patch.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCV.td [hunk 1]
```diff
@@ -54,6 +54,7 @@ include "RISCVSchedSyntacoreSCR1.td"
 include "RISCVSchedSyntacoreSCR345.td"
 include "RISCVSchedSyntacoreSCR7.td"
 include "RISCVSchedXiangShanNanHu.td"
+include "RISCVSchedTTAscalonD8.td"
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCV.td:57]: Alphabetize?

File: llvm/lib/Target/RISCV/RISCVSchedTTAscalonD8.td [hunk 2]
```diff
@@ -0,0 +1,333 @@
+//=- RISCVSchedTTAscalonD8.td - Tenstorrent Ascalon Scheduling Defs -----*- tablegen -*-=//
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVSchedTTAscalonD8.td:1]: Can we keep this at 80 characters?

File: llvm/lib/Target/RISCV/RISCVSchedTTAscalonD8.td [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+def : WriteRes<WriteRotateImm32, [AscalonFX]>;
+def : WriteRes<WriteRotateReg, [AscalonFX]>;
+def : WriteRes<WriteRotateReg32, [AscalonFX]>;
+
+def : WriteRes<WriteCLZ, [AscalonFX]>;
+def : WriteRes<WriteCLZ32, [AscalonFX]>;
+def : WriteRes<WriteCTZ, [AscalonFX]>;
+def : WriteRes<WriteCTZ32, [AscalonFX]>;
+
+def : WriteRes<WriteCPOP, [AscalonFX]>;
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVSchedTTAscalonD8.td:88]: I'm a little surprised CPOP is 1 cycle and on every ALU.

Comment (ppenzin) [reply llvm/lib/Target/RISCV/RISCVSchedTTAscalonD8.td:88]: I'll double check on that.

File: llvm/lib/Target/RISCV/RISCVSchedTTAscalonD8.td (hunk 2 above)
Comment (ppenzin) [reply llvm/lib/Target/RISCV/RISCVSchedTTAscalonD8.td]: Addressed.

File: llvm/lib/Target/RISCV/RISCV.td (hunk 1 above)
Comment (ppenzin) [reply llvm/lib/Target/RISCV/RISCV.td:56]: Addressed.

File: llvm/lib/Target/RISCV/RISCVSchedTTAscalonD8.td [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+}
+
+let SchedModel = TTAscalonD8Model in {
+
+//===----------------------------------------------------------------------===//
+// Define each kind of processor resource and number available.
+
+let BufferSize = 16 in {
+  def AscalonLS : ProcResource<3>;
+  def AscalonFXA : ProcResource<1>; // ALU, FP/VEC -> INT, MUL, DIV, CSR
```
Comment (mshockwave) [inline llvm/lib/Target/RISCV/RISCVSchedTTAscalonD8.td:36]: I noticed that in the uArch diagram, DIV and MUL are in different pipes. Is there any reason it's merged here?

File: llvm/lib/Target/RISCV/RISCVSchedTTAscalonD8.td [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+                             HasStdExtZkr, HasVInstructions, HasVInstructionsI64];
+}
+
+let SchedModel = TTAscalonD8Model in {
+
+//===----------------------------------------------------------------------===//
+// Define each kind of processor resource and number available.
+
+let BufferSize = 16 in {
+  def AscalonLS : ProcResource<3>;
```
Comment (mshockwave) [inline llvm/lib/Target/RISCV/RISCVSchedTTAscalonD8.td:35]: Can all three pipes run loads and stores? The uArch diagram only shows two load pipes but didn't mention the number of store pipe(s)

File: llvm/lib/Target/RISCV/RISCVSchedTTAscalonD8.td (hunk 4 above)
Comment (ppenzin) [reply llvm/lib/Target/RISCV/RISCVSchedTTAscalonD8.td:36]: You are right, the diagram doesn't reflect this definition, let me make sure it does going forward. I had to zoom into it in a separate browser tab to see the labels - never thought to check that before :)

File: llvm/lib/Target/RISCV/RISCVSchedTTAscalonD8.td (hunk 5 above)
Comment (ppenzin) [reply llvm/lib/Target/RISCV/RISCVSchedTTAscalonD8.td:35]: Another diagram issue noted, will follow up.

File: llvm/lib/Target/RISCV/RISCVSchedTTAscalonD8.td [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+//===----------------------------------------------------------------------===//
+
+def TTAscalonD8Model : SchedMachineModel {
+  let IssueWidth        =   8; // 8-way decode and dispatch
+  let MicroOpBufferSize = 256; // 256 micro-op re-order buffer
+  let LoadLatency       =   4; // Optimistic load latency
+  let MispredictPenalty =  14; // Fetch + Decode/Rename/Dispatch + Branch
+
+  // Enable partial & runtime unrolling.
+  let LoopMicroOpBufferSize = 16;
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVSchedTTAscalonD8.td:18]: If what you want is just tuning unrolling, then you don't need to override `LoopMicroOpBufferSize` here. You should add `TuneNoDefaultUnroll` to processor definition (which has already been done for `tt-ascalon-d8`, so you can just remove this line).

File: llvm/lib/Target/RISCV/RISCVSchedTTAscalonD8.td [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+// Define each kind of processor resource and number available.
+
+let BufferSize = 16 in {
+  def AscalonLS : ProcResource<3>;
+  def AscalonFXA : ProcResource<1>; // ALU, FP/VEC -> INT, MUL, DIV, CSR
+  def AscalonFXB : ProcResource<1>; // ALU, INT -> FP/VEC
+  def AscalonFXC : ProcResource<2>; // ALU, BR
+  def AscalonFXD : ProcResource<2>; // ALU
+  def AscalonFP : ProcResource<2>;
+  def AscalonV : ProcResource<2>;
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVSchedTTAscalonD8.td:41]: Vector units are not used. You may add it when adding vector scheduling support.

File: llvm/lib/Target/RISCV/RISCVSchedTTAscalonD8.td (hunk 6 above)
Comment (ppenzin) [reply llvm/lib/Target/RISCV/RISCVSchedTTAscalonD8.td]: Good point, addressed.

File: llvm/lib/Target/RISCV/RISCVSchedTTAscalonD8.td (hunk 7 above)
Comment (ppenzin) [reply llvm/lib/Target/RISCV/RISCVSchedTTAscalonD8.td]: Addressed.

File: llvm/lib/Target/RISCV/RISCVSchedTTAscalonD8.td (hunk 4 above)
Comment (ppenzin) [reply llvm/lib/Target/RISCV/RISCVSchedTTAscalonD8.td:33]: I can confirm that the diagram is not up to date, while this reflects the model we've been using internally for some time.

File: llvm/lib/Target/RISCV/RISCVSchedTTAscalonD8.td (hunk 5 above)
Comment (ppenzin) [reply llvm/lib/Target/RISCV/RISCVSchedTTAscalonD8.td:32]: Same here, I am pretty sure the PR is accurate, while diagram is not.

**Comment by ppenzin** [comment]
To the issues @mshockwave raised with microarch diagram on our website - I am trying to follow up on what is the backstory behind that. However, this PR is the definition we are using internally, and pretty confident it is accurate.

File: llvm/lib/Target/RISCV/RISCVSchedTTAscalonD8.td [hunk 8]
```diff
@@ -0,0 +1,330 @@
...
+
+def TTAscalonD8Model : SchedMachineModel {
+  let IssueWidth        =   8; // 8-way decode and dispatch
+  let MicroOpBufferSize = 256; // 256 micro-op re-order buffer
+  let LoadLatency       =   4; // Optimistic load latency
+  let MispredictPenalty =  14; // Fetch + Decode/Rename/Dispatch + Branch
+
+  let CompleteModel = 0;
+
+  // TODO supported, but haven't added scheduling info yet
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVSchedTTAscalonD8.td:19]:
```suggestion
  // TODO: supported, but haven't added scheduling info yet.
```

File: llvm/lib/Target/RISCV/RISCVSchedTTAscalonD8.td [hunk 9]
```diff
@@ -0,0 +1,330 @@
...
+// Define each kind of processor resource and number available.
+
+let BufferSize = 16 in {
+  def AscalonLS : ProcResource<3>;
+  def AscalonFXA : ProcResource<1>; // ALU, FP/VEC -> INT, MUL, DIV, CSR
+  def AscalonFXB : ProcResource<1>; // ALU, INT -> FP/VEC
+  def AscalonFXC : ProcResource<2>; // ALU, BR
+  def AscalonFXD : ProcResource<2>; // ALU
+  def AscalonFP : ProcResource<2>;
+  // TODO two vector units with vector scheduling model
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVSchedTTAscalonD8.td:38]:
```suggestion
  // TODO: two vector units with vector scheduling model.
```

File: llvm/lib/Target/RISCV/RISCVSchedTTAscalonD8.td [hunk 10]
```diff
@@ -0,0 +1,330 @@
...
+def : WriteRes<WriteIMul, [AscalonFXA]>;
+def : WriteRes<WriteIMul32, [AscalonFXA]>;
+}
+
+// Integer division
+// Worst case latency is used.
+
+let Latency = 7, ReleaseAtCycles = [7] in {
+  def : WriteRes<WriteIDiv32, [AscalonFXA]>;
+  def : WriteRes<WriteIDiv, [AscalonFXA]>;
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVSchedTTAscalonD8.td:69]: Is it right that 32/64 bits divisions have the same upper bound?

Comment (ppenzin) [reply llvm/lib/Target/RISCV/RISCVSchedTTAscalonD8.td:69]: Yes, they do.

**Review by wangpc-pp — APPROVED** [review]
LGTM.

**Review by mshockwave — APPROVED** [review]
LGTM

**Comment by ppenzin** [comment]
I need help merging the change, thanks in advance!

**Pull request merged**